File: leet_BTREE_0872_same_leaf_trees___DFS.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn leaf_similar (
        root1: Option<Rc<RefCell<TreeNode>>>, 
        root2: Option<Rc<RefCell<TreeNode>>>
    ) -> bool {
        fn dfs_on_single_node
        (
            node: &Option<Rc<RefCell<TreeNode>>>, 
            v: &mut Vec<i32>
        )
        {
            if let Some(ptr) = node
            {
                let ptr = ptr.borrow();
                if ptr.left.is_none() && ptr.right.is_none()
                {
                    v.push(ptr.val);
                    return ;
                }
                dfs_on_single_node( &ptr.left, v);
                dfs_on_single_node( &ptr.right, v);
            }
        }
        let mut a1 = Vec::new();
        let mut a2 = Vec::new();
        dfs_on_single_node( &root1, &mut a1);
        dfs_on_single_node( &root2, &mut a2);
        a1 == a2
    }
}

File: leet_BTREE_0872_same_leaf_trees___DFS.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */

var leafSimilar = function(root1, root2) {
    let dfs_on_single_node = (node, arr) => {
        if (!node) {
            return
        }
        if (!node.left && !node.right) {
            arr.push(node.val)
            return
        }
        dfs_on_single_node(node.left, arr)
        dfs_on_single_node(node.right, arr)
    }
    let a1 = []
    let a2 = []
    dfs_on_single_node(root1, a1)
    dfs_on_single_node(root2, a2)
    if (a1.length !== a2.length) {
        return false
    }
    let i = -1
    while (++i < a1.length) {
        if (a1[i] !== a2[i]) {
            return false
        }
    }
    return true
};
